Expvar integration and JSON stdout

jleeh opened the pull request:
For [#159707176](https://www.pivotaltracker.com/story/show/159707176).

Expvar integration that opens up the endpoint `/debug/vars` behind standard session authentication.

~~Implementation includes job spec metrics like previous PR's, but included as part of expvars output. The job spec metrics are collected on-boot and stored in RAM, removing any load on bolt when the endpoint is being scraped.~~

~~This doesn't yet include the functionality to update job metrics during service operation, it only will add new jobs but not their runs. Due to the complexity of adding new runs during operation, and the size of this PR already, it will be done after the acceptance of this.~~

No specific node metrics are now exported via expvar. Instead a `JSON_STDOUT` param has been added that configures the node to output all logs in JSON for easier parsing and shipping to ELK.

dimroc commented on services/application.go:
```diff
@@ -106,6 +109,7 @@ func (app *ChainlinkApplication) AddJob(job models.JobSpec) error {
 	}
 
 	app.Scheduler.AddJob(job)
+	app.JobMetrics.Add(job)
```
Should we have error handling?

dimroc commented on services/job_metrics.go:
```diff
@@ -0,0 +1,129 @@
+	var merr error
+
+	aj, err := jm.store.Jobs()
+	if err != nil {
```
This might bog down the system, having to iterate over every single job run on boot. Storing every job in memory with metrics could be ok, at least we're not storing every job run in memory.

dimroc replied:
A way around this would be to write and store these aggregate metrics as part of our normal operation `SaveRun`/`SaveJob` methods. Caching on write rather than calculating on boot.

dimroc replied:
Right 👍 

We also need to figure out the viability of holding every job in memory. Can a production chainlink node hold N jobs in memory? Maybe. JobRuns will grow far faster. Just something to think about it.

dimroc replied:
@jleeh I think this raises an interesting question: do you actually need all jobs?

Could we return the last 100 and allow you to paginate if you need more?

This way, we could write the metrics as they occur, and then simple get them from bolt when a GET reaches this endpoint. This wouldn’t be supported out the box by ‘expvar’ which is a bummer. I’m enjoying expvar but it seems better suited for global counts of constant size rather than an O(N) map of jobs.

dimroc commented on services/job_metrics.go:
```diff
@@ -0,0 +1,129 @@
+	tp := store.Config.MetricsParam
+
+	for _, p := range tp {
+		trp := jobRun.TaskRuns[0].Task.Params.Get(p)
```
Could you elaborate as to why we only check `TaskRuns[0]`? I have an idea but want to make sure.

jleeh replied:
Just due to run params being persisted through every task. This has changed now hasn't it so I should iterate and count through each task?

dimroc commented:
Hi there @jleeh, we discussed storing N job metrics in `expvar` and came to the following conclusions:

1. We hope to set a precedent with this approach that can be applicable to job runs, which has a much faster growth rate.
2. `expvar` is great for reporting metrics and aggregates that have an 0(K) memory growth rate, not great for reporting per job metrics (O(N)).
3. While Chainlink aspires to have what you need to run out the box, we don't think this includes per job/run metrics. We're of the mindset that we should feed this into another system.
4. Logs-based metrics is a prevalent practice that suits well for this scenario: using logs to report per N metrics. An example of this is the [Heroku log-runtime-metrics](https://devcenter.heroku.com/articles/log-runtime-metrics) or [Stackdriver logs-based-metrics](https://cloud.google.com/logging/docs/logs-based-metrics/) that has a separate service consume logs allowing arbitrary aggregation and filtering.

From this, we have two new action items:

1. Create top level job and job run aggregate stats for `expvar`: `jobs_created`, `job_runs_created`, `job_runs_errored`, `job_runs_completed`, `job_runs_in_progress`, `job_runs_pending`.
2. Introduce logs-based metrics for the N jobs or M job runs by placing key `logger.Metric` statements at key events: `job_created`, `job_run_created`, etc.

We'll be fleshing these out more offline and will be sure to include you. Could you list the metrics you're most interested in so we can be sure they're implemented @jleeh ?

jleeh commented:
Sounds perfect to me @dimroc, I already have an ELK stack that I'm pumping the logs into so it'd be very minimal changes for me to utilise log based metrics.

Would it be even be worth having any form of job metrics outputted in expvar? The ones you've mentioned you could probably just calculate using log metrics, rather leaving expvar to be completely standard and just output the utilisation metrics.

dimroc commented:
Excellent to hear you already have ELK. This will be a perfect trial then. As for `expvar`, you're right, there really isn't that much of a need at the moment for it. We can lower the priority of that or skip it all together.

Would still be great @jleeh to have a list of metrics you're interested in for this immediate story. Guess we can deduce them from this PR. Is it also an issue that you will only have the metrics when you are listening to the logs? i.e. if your ELK stack is down, you'll miss out on the reporting.

jleeh commented:
Personally, my most wanted metrics are:
- Number of jobs
- Numbers of job runs per jobs
- Adaptors/bridgers per job
- Run status changes
- Param used in runs
With outputting logging events on job creation, new run, and run updates, I should be able to construct the level of graphing as I have now.

Don't see it being much of an issue, I'm using Lambda to stream our logs to ELK currently and with relying on ELK for metrics, I'll up the amount of instances and add enough dedicated masters to ensure they can reach quorum if any go down.

jleeh commented:
As per discussion, removed the job metrics service and aggregates from the PR, rather going with the approach of gathering metrics by shipping logs over to ELK. I've added a `JSON_STDOUT` config param that is defaulted to false, but when true removes the pretty output from `log.jsonl` in the root dir.

se3000 commented on web/api.go:
```diff
@@ -51,8 +51,8 @@ func ParsePaginatedRequest(sizeParam, pageParam string) (int, int, int, error) {
-	query.Add("size", strconv.Itoa(size))
-	query.Add("page", strconv.Itoa(page))
+	query.Set("size", strconv.Itoa(size))
+	query.Set("page", strconv.Itoa(page))
```
👍
